### PR TITLE
[tune] Print warning msg when TrialExecutor is directly inherited.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

https://github.com/ray-project/ray/issues/17593

## Related issue number

#17593

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Yard1** [comment]
We need to merge https://github.com/ray-project/ray/pull/17584/files first, as there will be a metaclass conflict.

File: python/ray/tune/trial_executor.py [hunk 1]
```diff
@@ -8,7 +8,22 @@
...
+# A warning is printed to inform users of TrialExecutor deprecation.
+class TrialExecutorMetaclass(type):
+    def __new__(mcls, name, bases, module, **kwargs):
+        if name not in ("RayTrialExecutor", "_MockTrialExecutor",
+                        "TrialExecutor") and "RayTrialExecutor" not in tuple(
+                            base.__name__ for base in bases):
+            logger.warning(
+                f"TrialExecutor interface is being deprecated. "
+                f"RFC: https://github.com/ray-project/ray/issues/17593. "
+                f"Please contact Ray Tune if any concern.")
```
Comment (richardliaw) [inline python/ray/tune/trial_executor.py:21]: Please reach out on the Ray Github if you have any concerns

File: python/ray/tune/trial_executor.py [hunk 2]
```diff
@@ -8,7 +8,22 @@
...
-class TrialExecutor:
+# Signals when a class is directly inherited from TrialExecutor.
+# A warning is printed to inform users of TrialExecutor deprecation.
+class TrialExecutorMetaclass(type):
+    def __new__(mcls, name, bases, module, **kwargs):
+        if name not in ("RayTrialExecutor", "_MockTrialExecutor",
+                        "TrialExecutor") and "RayTrialExecutor" not in tuple(
+                            base.__name__ for base in bases):
+            logger.warning(
+                f"TrialExecutor interface is being deprecated. "
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/trial_executor.py:19]: how about warnings.warn(msg, DeprecationWarning)

[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/trial_executor.py]: Sounds good. Seems consistent with other warnings.warn usage in the codebase.

File: python/ray/tune/tests/test_no_direct_inherit_from_trial_executor.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                   f"RFC: https://github.com/ray-project/ray/issues/17593. "
+                   f"Please contact Ray Tune if any concern.")
+            assert ctx.records[0].msg == msg
+
+    def test_indirect_inheritance_is_ok(self):
+        with patch.object(logger, "warning") as warning_method:
+            from ray.tune.ray_trial_executor import RayTrialExecutor
+
+            class _MyRayTrialExecutor(RayTrialExecutor):
+                pass
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/tests/test_no_direct_inherit_from_trial_executor.py]: good catch. Updated!

**Comment by krfricke** [comment]
There seems to be a conflict, can you resolve this? Otherwise looks good to me.

Just FYI, there was a recent request about implementing an own trial executor: https://discuss.ray.io/t/integrate-my-own-executer-with-tune/3740/2

I still think this is an edge case so would like to go ahead, but maybe we can get more info from this user on what they are trying to exactly/what their use-case is.

**Review by krfricke — APPROVED** [review]
Looks good!

**Pull request merged**